src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -69,25 +72,73 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
             IThreadingContext threadingContext,
             VisualStudioWorkspaceImpl workspace,
             SVsServiceProvider serviceProvider,
-            IVsEditorAdaptersFactoryService editorAdaptersFactoryService)
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            [Import(AllowDefault = true)] Lazy<IVsPackageInstallerServices> packageInstallerServices,
+            [Import(AllowDefault = true)] Lazy<IVsPackageInstaller2> packageInstaller,
+            [Import(AllowDefault = true)] Lazy<IVsPackageUninstaller> packageUninstaller,
+            [Import(AllowDefault = true)] Lazy<IVsPackageSourceProvider> packageSourceProvider)
```

Comment:
I am not sure what lazy buy us here? isn't this called right away when StartWorking get called? right now issue was GetPackageSourncesAsync being called?

Reply:
by the way, isn't this whole thing already lazy through AbstractDelayStartedService (http://index/?leftProject=Microsoft.VisualStudio.LanguageServices&leftSymbol=rvtdcrtpzz7k&file=SymbolSearch%5CAbstractDelayStartedService.cs&line=21)

Reply:
okay, understood now why this is lazy. then assuming all nuget API can be called from BG.

not sure why 
http://index/?leftProject=Microsoft.VisualStudio.LanguageServices&leftSymbol=xyflomgaagny&file=SymbolSearch%5CAbstractDelayStartedService.cs&line=53

trying to make sure it is working from UI thread. we could put push all work to BG -
http://index/?leftProject=Microsoft.VisualStudio.LanguageServices&leftSymbol=ujz4ei5xsq7x&file=Implementation%5CLanguageService%5CAbstractPackage%602.cs&line=107

Reply:
feels like just moving call to Connect to BG solve all issues then?

Reply:
@heejaechang Moving the call to background is insufficient. The call was eliminated in its entirety.

Reply:
but the other one still runs from UI thread? why not making those to run from BG as well? those 2 Connect calling from UI thread seems not necessary at all?